gtdynamics/universal_robot/Joint.h

```diff
@@ -416,7 +416,9 @@ class Joint : public boost::enable_shared_from_this<Joint> {
    * @brief Create expression that constraint the pose of two links imposed by
    * the joint angle.
    */
-  gtsam::Vector6_ poseConstraint(uint64_t t = 0) const;
+  gtsam::Vector6_ poseConstraint(
+      uint64_t t = 0, const boost::optional<DynamicsSymbol> &wTp = boost::none,
+      const boost::optional<DynamicsSymbol> &wTc = boost::none) const;
```

Comment:
I think this might work better as an overloaded function since supplying t vs (wTp, wTc) should, I think, be mutually exclusive.

And presumably the non-t version should also be forced to specify the joint angle q.  Then that would revert the changes to `DynamicsGraph`, `KinematicsSlice`, PoseFactor line 66, and testPoseFactor.

Reply:
What is the reasoning for the constructors to be mutually exclusive? There would be code duplication in the overloaded method since the key is needed for the expression and I wanted to minimize the change.

Also going back to your prior comment about keys of the joint's link, that is something I found strange here. Why are we specifying the joint key when we have the joint? I can add the key back but then we would have to add a similar check for the joint key and use the specified one if it is not null.

Reply:
Correction: I will add the JointAngleKey back. :)

Reply:
Re: code duplication, I was thinking something like:

```c++
Vector6_ poseConstraint(const DynamicsSymbol &wTp,
                        const DynamicsSymbol &wTc,
                        const DynamicsSymbol &q) const {
  ...
}
Vector6_ poseConstraint(uint64_t t) const {
  return poseConstraint(PoseKey(parent->id(), t), PoseKey(child->id(), t), JointAngleKey(id(), t));
}
```

As for reasoning, it seemed weird to me to mix-and-match custom vs default keys.  As it stands, you can't specify a custom joint angle which seems inconsistent with the nature of the change, and if you could, then there are cases where the t argument is completely unused (i.e. when all 3 custom keys are specified).  But idk I guess 

>Also going back to your prior comment about keys of the joint's link, that is something I found strange here. Why are we specifying the joint key when we have the joint? I can add the key back but then we would have to add a similar check for the joint key and use the specified one if it is not null.

I'm not sure I follow this, but if you're referring to why PoseFactor constructor originally had 3 keys passed in, I am confused by that (as well).  If you're referring to why _I_ think the joint angle q should also be specified, I'd say for the same reason that you want the link pose keys specified.  "why are we specifying the joint key when we have the joint?" can be stated the same way for the pose key: "why are we specifying the link pose key when we have the joint (which has pointers to the links)?"  Mainly I guess I'm just thinking to be consistent.

Reply:
I agree with Gerry's way of redesigning the function signatures.

But let's merge this first since it's related to the code on quadruped state estimation project, and I'll make a future PR to address this.

Reply:
> As it stands, you can't specify a custom joint angle which seems inconsistent with the nature of the change, and if you could, then there are cases where the t argument is completely unused (i.e. when all 3 custom keys are specified). But idk I guess

I am not sure I follow. I thought the idea was to specify a custom joint angle via a prior on the joint angle variable. The thought behind this change was that we always set the joint angle variable to be a function of the joint ID and the time `t` hence providing a custom joint angle key didn't make sense?

> I'm not sure I follow this, but if you're referring to why PoseFactor constructor originally had 3 keys passed in, I am confused by that (as well). If you're referring to why I think the joint angle q should also be specified, I'd say for the same reason that you want the link pose keys specified. "why are we specifying the joint key when we have the joint?" can be stated the same way for the pose key: "why are we specifying the link pose key when we have the joint (which has pointers to the links)?" Mainly I guess I'm just thinking to be consistent.

I see your point about consistency here. Changing the link key made sense since link's are the primary objects that have mass and pose so they are the ones we care about.